Replace the Aggregate Checker with the Composite Checker

smillst opened the pull request:
* Moves the code that handles subcheckers into `SourceChecker`.
* Adds a new class called `CompositeChecker` that is a subclass of SourceChecker.  It is a replacement for the `AggregateChecker`. (Though, unlike the Aggregate Checker,  it interleaves the subcheckers error messages.) [[Should we just keep the Aggregate Checker name]]
* Deprecates the Aggregate Checker.

mernst reviewed (changes requested):
Could you please update the documentation in the manual as well?

https://checkerframework.org/manual/#creating-bundling-multiple-checkers talks of "aggregate checkers" and "compound checkers".  How does this fit in?  We have a chance to improve terminology and make it consistent between the documentation and the implementation.

smillst commented:
> Could you please update the documentation in the manual as well?
> 
> https://checkerframework.org/manual/#creating-bundling-multiple-checkers talks of "aggregate checkers" and "compound checkers". 

> How does this fit in? 

I answered this in the pull request description:
 
> * Adds a new class called `CompositeChecker` that is a subclass of SourceChecker.  It is a replacement for the `AggregateChecker`. (Though, unlike the Aggregate Checker,  it interleaves the subcheckers error messages.)  [[Should we just keep the Aggregate Checker name?]]

I will update the documentation once we decided that it makes sense to rename the Aggregate Checker to the Composite Checker.  Other than the rename, the only thing that needs to be changed about the manual is that the Aggregate checker interleaves the subcheckers error messages. (Also, as I noted in the pull request description, this needs a change log entry.)

mernst commented:
Can you please update the manual?  I would like to read the documentation before the code.

From a comment [comment](https://github.com/typetools/checker-framework/pull/6742?w=1#issuecomment-2275988520) that refers to text that is not in the pull request description, the pull request seems to replace AggregateChecker by CompositeChecker, but the title says "implement the Aggregate Checker".

What is the motivation to rename "Aggregate Checker" to "Compound Checker"?  I'm a bit concerned that "Composite Checker" and "Compound Checker" are exteremely similar names, but will refer to rather different ways to combine checkers.

smillst commented:
> Can you please update the manual? I would like to read the documentation before the code.
> 
> From a comment [comment](https://github.com/typetools/checker-framework/pull/6742?w=1#issuecomment-2275988520) that refers to text that is not in the pull request description, the pull request seems to replace AggregateChecker by CompositeChecker, but the title says "implement the Aggregate Checker".
> 
> What is the motivation to rename "Aggregate Checker" to "Compound Checker"? I'm a bit concerned that "Composite Checker" and "Compound Checker" are exteremely similar names, but will refer to rather different ways to combine checkers.

I feel like you are looking at stale version of this pull request. Originally this pull request added a Composite Checker, because that's what the code in https://github.com/typetools/checker-framework/pull/6723 did.  We discussed this and decided not to add the Composite Checker. Because it's a confusing name. Instead, I just changed the implementation of the Aggregate Checker to use the same implementation of compound checker. To do this I moved code from the BaseTypeChecker to SourceChecker so that AggregateChecker can use it. Then I did update the manual, but not much changed for the user point of view because I only really change how the Aggregate Checker was implement.  The only user visible changes to the Aggregate Checker are interleaved error messages and a different method needs to be overridden.

mernst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1102,6 +1362,11 @@ public void typeProcess(TypeElement e, TreePath p) {
+      printStoredMessages(p.getCompilationUnit());
+      if (!getSubcheckers().isEmpty()) {
+        // Update errsOnLastExit to reflect the errors issued.
+        this.errsOnLastExit = log.nerrors;
```
Why is it OK to update only one `errsOnLastExit` here?  Previous code set it for every checker including subcheckers.

smillst replied:
The code before did not update on every checker including subcheckers.  This code is copied from `BaseTypeChecker#typeProcess` in two chunks. The first chunk is the part above the call to super.typeProcess which was copied to the top of this method: https://github.com/typetools/checker-framework/blob/723ec826cfbce855b395ed3228f5ea4df4bc720b/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java#L527-L558

The second chunk after the call to super.typeProcess was copied to the end of this method:
https://github.com/typetools/checker-framework/blob/723ec826cfbce855b395ed3228f5ea4df4bc720b/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java#L559-L565

Every subchecker eventually ends up setting`this.errsOnLastExit` when typeProcess is called on it.

mernst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1676,6 +1965,77 @@ protected final void setLintOption(String name, boolean val) {
   }
 
+  /**
+   * Returns the set of subchecker classes on which this checker depends. Returns an empty set if
```
What does it mean to "depend"?  Is this different than being a subchecker?

smillst replied:
It's a subchecker of this current checker rather than a subchecker of some other checker.

mernst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1676,6 +1965,77 @@ protected final void setLintOption(String name, boolean val) {
+   * call the super implementation of this method and add dependencies to the returned set so that
+   * checkers required for reflection resolution are included if reflection resolution is requested.
+   *
+   * <p>If you wish to create a checker that runs multiple checkers, but itself is not a type
```
Is this line about running multiple independent checkers, or multiple checkers that may depend on one another?

mernst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1676,6 +1965,77 @@ protected final void setLintOption(String name, boolean val) {
+   * <p>This method is protected so it can be overridden, but it should only be called internally by
+   * {@link SourceChecker}.
+   *
+   * @return the subchecker classes on which this checker depends; will be modified by callees
+   */
+  // This is never looked up in, but it is iterated over (and added to, which does a lookup).
+  protected Set<Class<? extends SourceChecker>> getImmediateSubcheckerClasses() {
```
Is the contrast between "depends" and "subchecker" intentional?  What difference is it emphasizing?

smillst replied:
Yes, this method is only dealing with subcheckers that this checker uses.  Any of those subchecker may also have subcheckers, but this method does not return those.

mernst replied:
Can we eliminate the new term "depends", so that we don't have to define it, and use "immediate subchecker" throughout instead?

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1689,33 +2049,47 @@ public Set<String> getSupportedLintOptions() {
+   *
+   * @return the set of supported lint options for this checker and its subcheckers
+   */
   protected Set<String> createSupportedLintOptions() {
```
BaseTypeChecker overrides it to add  lint options like "cast".

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1905,32 +2257,103 @@ public final List<String> getStringsOption(
+
+    int msgcmp = o1.message.compareTo(o2.message);
+    if (msgcmp == 0) {
+      // If the two messages are identical so far, it doesn't matter
```
This code was copied from `BaseTypeChecker`.  I think you're right that is about deduplicating messages, but I'm not sure.

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1003,10 +1124,120 @@ public void initChecker() {
+      instance.setProcessingEnvironment(this.processingEnv);
+      instance.treePathCacher = this.getTreePathCacher();
+      // Prevent the new checker from storing non-immediate subcheckers
+      instance.subcheckers = Collections.emptyList();
+      immediateSubcheckers.add(instance);
+      instance.immediateSubcheckers =
+          instance.instantiateSubcheckers(alreadyInitializedSubcheckerMap);
+      instance.setParentChecker(this);
+      alreadyInitializedSubcheckerMap.put(subcheckerClass, instance);
```
I didn't write this code.  I think the order is important, but `immediateSubcheckers.add(instance);,` could be first.  I've made that change, hopefully nothing breaks.

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1015,6 +1246,37 @@ public void initChecker() {
+    // All other messages are printed immediately.  This includes errors issued because the
+    // checker threw an exception.
+
+    // In order to run the next checker on this compilation unit even if the previous issued
```
No, it means what ever checker is next to be run.  The next checker could be "the checker for which the current checker is a subchecker", but it also could be an unrelated checker in an aggregate checker.

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -758,6 +822,37 @@ public Trees getTreeUtils() {
 
+  /**
+   * Like {@link #getOptions}, but only includes options passed to this checker. Does not include
+   * those passed to subcheckers.
```
This is true.  But I feel like rewording the documentation would make it less clear.

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -1905,32 +2257,102 @@ public final List<String> getStringsOption(
+    if (checkercmp != 0) {
+      return checkercmp;
+    }
+    // The two messages are from the same checker, so sort by message.
```
Yes, I think we may want to look into this method a tweak it in another pull request.

mernst replied:
@smillst Can you please open an issue or a pull request regarding this?  Thanks.

smillst commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -601,7 +634,36 @@ public abstract class SourceChecker extends AbstractTypeProcessor implements Opt
+
+  /**
+   * The full list of subcheckers that need to be run prior to this one, in the order they need to
+   * be run in. This list will only be non-empty for the one checker that runs all other
```
It can only be null or an immutable list which is empty for all but the ultimate parent checker. (I updated the documentation.) Can you point out where it can be a mutable list?

mernst commented on framework/src/main/java/org/checkerframework/framework/source/AggregateChecker.java:
```diff
@@ -23,6 +23,30 @@ public abstract class AggregateChecker extends SourceChecker {
+   * @return the independent subchecker classes; will be modified by callees
+   */
   @Override
   protected abstract Set<Class<? extends SourceChecker>> getImmediateSubcheckerClasses();
```
There is largely duplicate, but different, documentation on this method in AggregateChecker and in SourceChecker.
Can one of them be reduced or eliminated, to avoid the likelihood that the documentation will get out of sync?

smillst replied:
(I removed the documentation.)

mernst commented on framework/src/main/java/org/checkerframework/framework/source/AggregateChecker.java:
```diff
@@ -23,6 +23,30 @@ public abstract class AggregateChecker extends SourceChecker {
   protected AggregateChecker() {}
 
+  /**
+   * Returns the set of independent subchecker classes run by this checker.
```
That is good.  Thanks.
